### ISSUE TITLE
Azure Storage Blob SDK Uploads Only 250MB

### ISSUE THREAD

**Issue by brt5586** [pglombardo/PasswordPusher#1208]
Issue when uploading to a storage blob, 250MB is the hard limit - this is easy to reproduce - currently got NGINX as the frontend proxy, postgres as docker container and same with pwpusher on Ubuntu 22.04 on Azure VM.

**The issue:**
When uploading to storage blob, you will notice in console '413' error response indicating the file size is too large. Looking further into this and it's to do with the SAS service version: (outlined here: https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=azure-ad#remarks)

![image](https://github.com/pglombardo/PasswordPusher/assets/22344093/8f698002-447e-4e9c-a33f-7e32ba48369a)

https://REDACTED.blob.core.windows.net/REDACTED/randomfile?sp=rw&sv=2018-11-09&se=2023-06-28T10%3A00%3A50Z&sr=REDACTED

As seen above it's requested the 2018-11-09 service version which is 250MB limit, since then it's now been upped 5GB with newer service versions.

I'm using the latest version of pwpusher, please can you change this?

**Comment by pglombardo** [comment]
Hi @brt5586 - thanks this is useful info. I'm traveling for the next couple weeks but I'll take a closer look and see if I can make the improvement once I get back.

**Comment by pglombardo** [comment]
Hi @brt5586 I previously didn't have this documented well but could you confirm that you put in the proper CORS configuration for the Azure storage?  One of the CORS configurations is to allow the `x-ms-blob-type` header.

Updated documentation for Azure is here: https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#azure-storage-cors-configuration

**Comment by msundman78** [comment]
I'm having the same problem, being limited to uploading max 256MB files to Azure Blob Storage due to the old API version being used. According to this thread:
https://github.com/Azure/azure-storage-ruby/issues/172

It should be possible to pass the following setting to force a newer version:
Azure::Storage::Blob::Default::STG_VERSION = "xxxx-xx-xx" 

I believe 2024-05-04 is the current version recommended to be used:
https://learn.microsoft.com/en-us/rest/api/storageservices/versioning-for-the-azure-storage-services

Is that anything you can use from PWP?

**Comment by pglombardo** [comment]
Hi @msundman78 - that's helpful thanks.  I dug in a little bit and Azure support doesn't seem to be in great shape.  I'll document here for the record.

Password Pusher uses [ActiveStorage](https://guides.rubyonrails.org/active_storage_overview.html) to upload/store/retrieve files.  [This is the section ](https://guides.rubyonrails.org/active_storage_overview.html#microsoft-azure-storage-service)on how to configure Azure storage.

It references [the azure-storage-blob Ruby gem](https://rubygems.org/gems/azure-storage-blob) which hasn't been updated since 2021. :-(

When I go to [the source code for that gem](https://github.com/Azure/azure-storage-ruby), it's been deprecated by Microsoft with no clear alternative.

This is the discussion in the Rails community: https://github.com/rails/rails/issues/49983

@msundman78 you _might_ be able to manually set that value.  You could create a file `config/initializers/azure.rb` with the following content:

```ruby
require "azure-storage-ruby"
Azure::Storage::Blob::Default::STG_VERSION = "2024-05-04"
```

**Comment by msundman78** [comment]
Thanks. Unfortunately it wasn't that easy. ActiveStorage is using SAS tokens, which does not use the x-ms-version header for selecting the API version. Instead the API version is selected when generating the SAS URL in the **SignedVersion** (sv) parameter, or the optional **api-version** parameter.

https://learn.microsoft.com/en-us/rest/api/storageservices/versioning-for-the-azure-storage-services#authorize-requests-by-using-a-shared-access-signature

I found that I could override the default SignedVersion version in the Common module instead of the Blob module with:
```
require 'azure/storage/blob'
module Azure::Storage::Common::Default
  remove_const(:STG_VERSION) if const_defined?(:STG_VERSION)
  const_set(:STG_VERSION, '2024-05-04')
end
```
But the SignedVersion is part of the SAS signature, and it seems like it is not creating the signature based on the new SV, as when I test it I can see the SV parameter is using the new version but the PUT requests fails with:

_403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature._

I tested with curl and generated a SAS URL in the Azure portal which gives me the same error if I just change the SV in the URL. However if I instead add api-version=2024-05-04 at the end of the URL (which is not part of the signature calculation) it successfully uses the new API version (Which is confirmed by the response headers saying x-ms-version=2024-05-04).

I need to dig a little deeper so see if I've missed something or if we need to patch ActiveStorage to solve it.

**Comment by msundman78** [comment]
Yeah, I just got it working! My assumption the SV parameter was not included in the signature calculation was wrong. I think the the problem was that the most recent API versions simply is not supported as SAS SignedVersions. I changed to version 2020-02-10 and then it worked! I just successfully pushed a 1.3GB file.

**Comment by msundman78** [comment]
According to this the signature format changed slightly from v2020-12-06 which explains why the newer versions doesn't work just out of the box:

https://learn.microsoft.com/en-us/rest/api/storageservices/create-service-sas#version-2020-12-06-and-later

**Comment by pglombardo** [comment]
@msundman78 - that is an excellent investigation and solution!  You should also post this in https://github.com/rails/rails/issues/49983 for others in the Rails community that are facing this.

For anyone else reading this thread and using Azure as a backend, could you confirm that the fix above works for you?  If so, I'll roll this initializer into the main code base.

Thanks @msundman78!

**Comment by msundman78** [comment]
@pglombardo,
Have you merged the above change, or are you still waiting for more people to comment? 

I also wonder how ActiveStorage decision to drop support for azure-storage-blob will affect Azure Blob support in pwpush?
https://github.com/rails/rails/pull/52863

If I understand this blogg right, it should be able to switch to Active Storage adapter to use the new azure-blob gem instead:
https://github.com/rails/rails/pull/52863

**Comment by pglombardo** [comment]
No... this issue just got lost in the noise and I haven't been back with all the other things happening.

That is a great find.  I'll be upgrading the app to Rails 8 soon - when I do I'll switch over to that new azure-blob gem.
